### ISSUE TITLE
Stutter before starting playing?

### ISSUE THREAD

**Issue by SpinEve** [Aleksoid1978/VideoRenderer#43]
Between opening video and before playing, there's a stutter like one second long. In this period the display stops updating (and also the pointer), which is a bit annoying.

It seems somehow related to the Intel + Nvidia GPU setting on laptop, because when I run MPC-BE with VR on dedicated GPU, the stutter is gone. Also, the stutter doesn't happen with MPC-BE & inbuilt EVR renderer.

It should not a problem of specific system, as I had exactly the same problem on my previous laptop. However I couldn't find any solutions and discussions on Internet :(

My MPC-BE version is 1.6.0.6767, and I tested with both VR 0.5.6.1807 and 0.5.7.1812.

**Comment by clsid2** [comment]
Test if disabling "Use Direct3D 11" solves it.

Which Intel and Nvidia GPU models?

**Comment by SpinEve** [comment]
No, it exists in both D3D9 and D3D11 mode.

I use a I9-11900H and RTX3060 model.

**Comment by Aleksoid1978** [comment]
Record video from desktop with stutter, enable statistics(Ctrl+J) in MPC VR.

**Comment by SpinEve** [comment]
I know it's weird, but if I run my OBS (even NOT start recording), the stutter magically disappears.

But I can give a screenshot of shutter (with no recording software running) here: https://imgur.com/a/AV55boj

**Comment by SpinEve** [comment]
The stutter is not always so obvious in the statistics, and the length seems to be shorter in D3D9 mode, but it exists in all situation.

**Comment by SpinEve** [comment]
Actually, I find that the stutter is gone as long as there's any application using dedicated (Nvidia) GPU at the same time, like games. 

It also works if I just completely disable Nvidia GPU in utility of my laptop. Related to GPU initialization I guess?

**Comment by clsid2** [comment]
You can set Nvidia (or Intel) as preferred GPU for MPC-BE in the Windows 10 settings.
Settings > Display > Advanced

**Comment by SpinEve** [comment]
Yes I can (and that's how I did in the first post), but it won't solve the stutter if Intel GPU is set as preferred. And for a laptop, I really don't want to use power-consuming Nvidia GPU just for video playing...

**Comment by clsid2** [comment]
Does it also happen with MPC-HC? I ask because the Nvidia driver has an application profile for MPC-HC, which defaults to the integrated GPU. I don't know if this overrules the Windows setting.

If it works better with MPC-HC, then you can add the BE exe names to its profile with Nvidia Profile Inspector.

**Comment by SpinEve** [comment]
> Does it also happen with MPC-HC? I ask because the Nvidia driver has an application profile for MPC-HC, which defaults to the integrated GPU. I don't know if this overrules the Windows setting.
> 
> If it works better with MPC-HC, then you can add the BE exe names to its profile with Nvidia Profile Inspector.

I did a quick test and unfortunately MPC-HC situation is basically the same. (with your fork of MPC-HC 1.9.18)

**Comment by Aleksoid1978** [comment]
Developers don't have such hardware and can't debug/test.

**Comment by Aleksoid1978** [comment]
It looks like the system or video driver, when starting playback, does some initialization when using Intel. And it takes a long time. There is no such code in the renderer.

**Comment by SpinEve** [comment]
> It looks like the system or video driver, when starting playback, does some initialization when using Intel. And it takes a long time. There is no such code in the renderer.

I agree. But EVR (and I remember madVR on my previous system) doesn't have this issue. I'll go back to EVR for now and maybe try to locate the problem afterwards. Thanks for your suggestions!

**Comment by clsid2** [comment]
EVR uses D3D9. It also does not use the DXVA2 video processor for scaling. So test if disabling the video processor in MPCVR makes any difference.
Open the Intel graphics control panel. In the video section, find the option "Contrast Enhancement" and disable that. This feature is useless, and known to cause a whole lot of problems.

@Aleksoid1978
A debug build plus DebugView tool could maybe show which part of the code has the delay?

**Comment by clsid2** [comment]
Are you using D3D11 hardware acceleration in the video decoder? Test with DXVA2 decoding.

There was a fix in LAV Filters today that solves a startup delay with D3D11 decoding.

**Comment by SpinEve** [comment]
> Are you using D3D11 hardware acceleration in the video decoder? Test with DXVA2 decoding.
> 
> There was a fix in LAV Filters today that solves a startup delay with D3D11 decoding.

I'm using MPC video decoder with MPC-BE, and all decoders (DXVA2, D3D11 and D3D11CB) has this problem, but NVDEC doesn't. I'll test the Intel driver and LAV filters later.

**Comment by SpinEve** [comment]
Ok I tried with latest LAV filter build (0.75.1.45-git) and its decoders. Result is similar: Only Nvidia CUVID decoder has no stutter, and even "None" (software decoder I guess) is problematic. 

Turning on/off Contrast Enhancement and other various options in Intel control panel also doesn't make any differences.

**Comment by Aleksoid1978** [comment]
As i see on screenshots, there is no issue when using D3D11 Native in decoder and D3D11 in VR.

**Comment by SpinEve** [comment]
> As i see on screenshots, there is no issue when using D3D11 Native in decoder and D3D11 in VR.

Yes, if I use D3D11, there's no issue in resizing, only in starting playing. It's hard to be shown in a screenshot, since it happens before rendering videos and statistics. 

But the situations are pretty similar to resizing issues in D3D9. Just wandering if the causes are related and help debugging.

**Comment by SpinEve** [comment]
https://user-images.githubusercontent.com/20128868/149972941-c2c7c937-9de6-494d-a392-d587cec062ac.mp4

Ok I use my phone to record the issue. You can see that before first start playing, the window of MPC-BE and the mouse (actually the whole screen) freezes for a while. It also happens when loading.

**Comment by Aleksoid1978** [comment]
Hmm, normal opening.

**Comment by SpinEve** [comment]
> Hmm, normal opening.

Well I forget to note that I'm moving my mouse all the time in the recording. It's not quite normal that when opening videos, the screen freezes and the cursor stops moving...

**Comment by Aleksoid1978** [comment]
Here is debug version + dbgview - https://disk.yandex.ru/d/0cFtmsMD81QQdA
Register this version or replace over you's.
Run dbgview, then open video in MPC-BE. Close MPC-BE and save log in dbgview. Then upload here, better in .zip format.

**Comment by Aleksoid1978** [comment]
If mouse cursor stop moving - it's something with you system, it's freeze. Software can't prevent mouse cursor from moving.

**Comment by SpinEve** [comment]
Here's the log: [mpc.zip](https://github.com/Aleksoid1978/VideoRenderer/files/7894040/mpc.zip) - but as you said, it seems not a problem of software.

After searching around the Internet, I found several posts about this "micro stutter" problem. The stutter is caused by dediacted GPU switching, especially when dGPU is awaked. I confimed it by installing an older Nvidia driver which keeps dGPU awaken more aggresively.

However, it's strange that neither the LAV decoder and MPCVR actually uses dGPU in this situation - they all on Intel GPU and dGPU is just awaken and went back to sleep for a while, without any workloads. If the GPU goes back to sleep, the stutter comes back if a new video starts playing (and wake up dGPU again, which cause the stutter).

So, if the switching stutter is problem of driver (or system), I wonder if it's possible to avoid waking up the dGPU in MPCVR, just like setting MPC-BE just on Intel GPU? Maybe an option in configuration?
